File: Backend/insurance-service/src/main/java/com/tpbs/insuranceservice/service/InsuranceService.java
package com.tpbs.insuranceservice.service;

import com.tpbs.insuranceservice.dto.InsuranceDto;

import java.util.List;
import java.util.Map;

public interface InsuranceService {
    
    List<InsuranceDto> getAllInsurance();
    InsuranceDto getInsuranceById(Long id);
    List<InsuranceDto> getInsuranceByUserId(Long userId);
    List<InsuranceDto> getInsuranceByBookingId(Long bookingId);
    InsuranceDto createInsurance(InsuranceDto insuranceDto);
    InsuranceDto updateInsurance(Long id, InsuranceDto insuranceDto);
    void deleteInsurance(Long id);
    InsuranceDto cancelInsurance(Long id);
    InsuranceDto renewInsurance(Long id);
    
    // Enhanced methods with cross-service data
    List<Map<String, Object>> getUserInsuranceWithBookingDetails(Long userId);
    Map<String, Object> getInsuranceWithBookingAndUserDetails(Long insuranceId);
    List<Map<String, Object>> getInsuranceQuotes(Long bookingId);
}

File: Frontend/src/services/insuranceService.js
import { apiRequest } from './apiClient';

class InsuranceService {
  constructor() {
    this.baseURL = '/api/insurance'; // API Gateway route
  }

  async getAllInsurance() {
    return await apiRequest(this.baseURL);
  }

  async getInsuranceById(insuranceId) {
    return await apiRequest(`${this.baseURL}/${insuranceId}`);
  }

  async getInsuranceByUserId(userId) {
    return await apiRequest(`${this.baseURL}/user/${userId}`);
  }

  async getInsuranceByBookingId(bookingId) {
    return await apiRequest(`${this.baseURL}/booking/${bookingId}`);
  }

  async createInsurance(insuranceData) {
    return await apiRequest(this.baseURL, {
      method: 'POST',
      data: insuranceData
    });
  }

  async updateInsurance(insuranceId, insuranceData) {
    return await apiRequest(`${this.baseURL}/${insuranceId}`, {
      method: 'PUT',
      data: insuranceData
    });
  }

  async cancelInsurance(insuranceId) {
    return await apiRequest(`${this.baseURL}/${insuranceId}/cancel`, {
      method: 'PUT'
    });
  }
  async getInsuranceTypes() {
    return await apiRequest(`${this.baseURL}/types`);
  }
  
  // Enhanced methods with cross-service data
  async getUserInsuranceWithDetails(userId) {
    return await apiRequest(`${this.baseURL}/user/${userId}/with-details`);
  }
  
  async getInsuranceWithDetails(insuranceId) {
    return await apiRequest(`${this.baseURL}/${insuranceId}/with-details`);
  }
  
  async getInsuranceQuotes(bookingId) {
    return await apiRequest(`${this.baseURL}/quotes/booking/${bookingId}`);
  }
  
  async renewInsurance(insuranceId) {
    return await apiRequest(`${this.baseURL}/${insuranceId}/renew`, {
      method: 'PUT'
    });
  }
}

export default new InsuranceService();